### PR TITLE
Fix URLs join failed when path contains `//`

### LINKED ISSUE #245: urllib.parse.urljoin use is wrong in case double slash path
Now m3u8
```python
>>> import os
>>> import posixpath
>>> from urllib.parse import urlparse, urljoin
>>>
>>> def _parsed_url(url):
...     parsed_url = urlparse(url)
...     prefix = parsed_url.scheme + '://' + parsed_url.netloc
...     base_path = posixpath.normpath(parsed_url.path + '/..')
...     return urljoin(prefix, base_path)
...
>>> def _urijoin(base_uri, path):
...     if base_uri.startswith('http'):
...         return urljoin(base_uri, path)
...     else:
...         return os.path.normpath(os.path.join(base_uri, path.strip('/')))
...
>>> # above codes just for view, here import real functoins
>>> from m3u8 import _parsed_url
>>> from m3u8.mixins import _urijoin
>>>
>>> # _parsed_url
>>> _parsed_url('http://hostname/path1/path2')  # normal path, only this is right
'http://hostname/path1'
>>> _parsed_url('http://hostname//path1/path2')  # double slash in front of the path
'http://path1'
>>> _parsed_url('http://hostname/path1//path2')  # double slash in middle of the path
'http://hostname/path1'
>>>
>>> # _urijoin
>>> _urijoin('http://hostname/path1/', 'path2')  # normal path, only this is right
'http://hostname/path1/path2'
>>> _urijoin('http://hostname//path1/', 'path2')  # double slash in front of the path
'http://hostname/path1/path2'
>>> _urijoin('http://hostname/path1//', 'path2')  # double slash in end of the path
'http://hostname/path1/path2'
>>> _urijoin('http://hostname/path1/', '//path2')  # double slash in front of the second path
'http://path2'
```
Changes
```python
>>> import os
>>> from urllib.parse import SplitResult, urlsplit
>>>
>>> def _parsed_url(url):
...     parsed_url = urlsplit(url)
...     base_path = parsed_url.path.rpartition('/')[0] + '/'
...     return SplitResult(parsed_url.scheme, parsed_url.netloc, base_path, '', '').geturl()
...
>>> def _urijoin(base_uri, path):
...     if path[0] == '/':
...         path = path[1:]
...     if base_uri.startswith('http'):
...         return base_uri + path
...     else:
...         return os.path.normpath(os.path.join(base_uri, path))
...
>>> # _parsed_url
>>> _parsed_url('http://hostname/path1/path2')  # normal path
'http://hostname/path1/'
>>> _parsed_url('http://hostname//path1/path2')  # double slash in front of the path
'http://hostname//path1/'
>>> _parsed_url('http://hostname/path1//path2')  # double slash in middle of the path
'http://hostname/path1//'
>>>
>>> # _urijoin
>>> _urijoin('http://hostname/path1/', 'path2')  # normal path
'http://hostname/path1/path2'
>>> _urijoin('http://hostname//path1/', 'path2')  # double slash in front of the path
'http://hostname//path1/path2'
>>> _urijoin('http://hostname/path1//', 'path2')  # double slash in end of the path
'http://hostname/path1//path2'
>>> _urijoin('http://hostname/path1/', '//path2')  # double slash in front of the second path
'http://hostname/path1//path2'
```

### PR DESCRIPTION
**Pull request by SeaHOH**
close #245

### CODE REVIEW

**Comment by SeaHOH** [comment]
@leandromoreira @mauricioabreu
Nobody gives a reveiw?

**Review by mauricioabreu — APPROVED** [review]

**Comment by mauricioabreu** [comment]
@SeaHOH does this happen when `/` is duplicated without any reason, right?
If this is the problem, maybe replacing `//` by `/` would be enough. What do you think?

**Comment by SeaHOH** [comment]
> does this happen when `/` is duplicated without any reason, right?

I had meet it a m3u8 url which contains `//`, so #245 has been submitted.

> If this is the problem, maybe replacing `//` by `/` would be enough. What do you think?

No, that could not resolve problem.

File: tests/test_loader.py [hunk 1]
```diff
@@ -74,13 +74,13 @@ def test_load_should_create_object_from_uri_with_relative_segments():
     prefix = urlparsed.scheme + '://' + urlparsed.netloc
     expected_key_abspath = '%s%s/key.bin' % (prefix, os.path.normpath(base_uri + '/..'))
     expected_key_path = '../key.bin'
-    expected_ts1_abspath = '%s/entire1.ts' % (prefix)
+    expected_ts1_abspath = '%s%s/entire1.ts' % (prefix, base_uri)
```
Comment (leandromoreira) [inline tests/test_loader.py:77]: if what we want to treat is when ta malformed manifest is introducing a double `//` I think we should create a new entire test, do you agree?

Comment (leandromoreira) [reply tests/test_loader.py:77]: or if we want to test legit URLs with `//` I think we should create a test demonstrating that, what do you think?

**Review by leandromoreira — COMMENTED** [review]
> > does this happen when `/` is duplicated without any reason, right?
> 
> I had meet it a m3u8 url which contains `//`, so #245 has been submitted.
> 
> > If this is the problem, maybe replacing `//` by `/` would be enough. What do you think?
> 
> No, that could not resolve problem.

Do you think that if we treat `http://hostname//path1/path2` as `http://hostname/path1/path2` wouldn't be better?

File: tests/test_loader.py (hunk 1 above)
Comment (SeaHOH) [reply tests/test_loader.py:77]: This is just a fix for new function result.

Of course new testcases may be needed, but notice that double means two or more if you add new.

**Comment by SeaHOH** [comment]
> Do you think that if we treat `http://hostname//path1/path2` as `http://hostname/path1/path2` wouldn't be better? or do you need to keep the double `//` ?

Yes, we need keep it as double or more.

File: tests/test_loader.py (hunk 1 above)
Comment (futzu) [reply tests/test_loader.py:77]: **Doesn't urljoin give you what you want?**

```python3
>>>> from urllib.parse import urljoin

>>>> baseurl="https://futzu.com//"
>>>> urljoin(baseurl,"/seg.ts")

'https://futzu.com/seg.ts'

>>>> baseurl="//"
>>>> urljoin(baseurl,"/seg.ts")

'/seg.ts'

>>>> urljoin(baseurl,"//seg.ts")

'//seg.ts'

>>>> urljoin(baseurl,"//seg.ts//")

'//seg.ts//'
```

Comment (SeaHOH) [reply tests/test_loader.py:77]: > Doesn't urljoin give you what you want?

It doesn't.
```python
>>> from urllib.parse import urljoin
>>> baseurl = "https://futzu.com/"
>>> urljoin(baseurl, "//seg.ts")
'https://seg.ts'
```

**Pull request closed without merging**